tls/s2n_encrypted_extensions.c

```diff
@@ -14,13 +14,118 @@
+  * certificates. 
+  **/
+
+static int s2n_server_encrypted_extensions_parse(struct s2n_connection *conn, struct s2n_blob *extensions);
 
 int s2n_encrypted_extensions_send(struct s2n_connection *conn)
 {
-    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+    struct s2n_stuffer *out = &conn->handshake.io;
+
+    /* Calculate size of encrypted extensions. For minimal TLS 1.3, this is 0
+     * as we are sending an empty EE message
+     */
+    uint16_t total_size = 0;
+
+    /* Write length of extensions */
+    GUARD(s2n_stuffer_write_uint16(out, total_size));
+
+    if (total_size == 0) {
+        return 0;
+    }
+
+    /* Write the extensions to the out buffer. For minimal TLS 1.3, this is
+     * a noop, as we are sending an empty EE message
+     */
+    
+    return 0;
 }
 
 int s2n_encrypted_extensions_recv(struct s2n_connection *conn)
 {
-    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+    struct s2n_stuffer *in = &conn->handshake.io;
+    uint16_t extensions_size;
+
+    /* Read encrypted extensions size */
+    S2N_ERROR_IF(2 > s2n_stuffer_data_available(in), S2N_ERR_BAD_MESSAGE);
+    GUARD(s2n_stuffer_read_uint16(in, &extensions_size));
+    S2N_ERROR_IF(extensions_size > s2n_stuffer_data_available(in), S2N_ERR_BAD_MESSAGE);
+
+    /* Process extensions */
+    if (extensions_size > 0) {
+        struct s2n_blob extensions = {0};
+        extensions.size = extensions_size;
+        extensions.data = s2n_stuffer_raw_read(in, extensions.size);
+        notnull_check(extensions.data);
+
+        GUARD(s2n_server_encrypted_extensions_parse(conn, &extensions));
+    }
+
+    return 0;
+}
+
+int s2n_server_encrypted_extensions_parse(struct s2n_connection *conn, struct s2n_blob *extensions)
+{
+    struct s2n_stuffer in = {0};
+
+    GUARD(s2n_stuffer_init(&in, extensions));
+    GUARD(s2n_stuffer_write(&in, extensions));
+
+    while (s2n_stuffer_data_available(&in)) {
+        struct s2n_blob ext = {0};
+        uint16_t extension_type, extension_size;
+        struct s2n_stuffer extension = {0};
+
+        GUARD(s2n_stuffer_read_uint16(&in, &extension_type));
+        GUARD(s2n_stuffer_read_uint16(&in, &extension_size));
+
+        ext.size = extension_size;
+        ext.data = s2n_stuffer_raw_read(&in, ext.size);
+        notnull_check(ext.data);
+
+        GUARD(s2n_stuffer_init(&extension, &ext));
+        GUARD(s2n_stuffer_write(&extension, &ext));
+
+        switch (extension_type) {
+        case TLS_EXTENSION_SERVER_NAME:
+            GUARD(s2n_recv_server_server_name(conn, &extension));
+            break;
+        case TLS_EXTENSION_ALPN:
+            GUARD(s2n_recv_server_alpn(conn, &extension));
+            break;
+        case TLS_EXTENSION_SCT_LIST:
+            GUARD(s2n_recv_server_sct_list(conn, &extension));
+            break;
+        case TLS_EXTENSION_MAX_FRAG_LEN:
+            GUARD(s2n_recv_server_max_frag_len(conn, &extension));
+            break;
+        /* Error on known extensions that are not supposed to appear in EE */
```

Comment:
perhaps add a link to https://tools.ietf.org/html/rfc8446#page-37

Reply:
will do